src/main.cpp

```diff
@@ -4049,9 +4051,11 @@ bool ProcessNewBlock(CValidationState& state, const CChainParams& chainparams, c
 
         // Store to disk
         CBlockIndex *pindex = NULL;
-        bool ret = AcceptBlock(*pblock, state, chainparams, &pindex, fRequested, dbp);
+        bool fNewBlock = false;
+        bool ret = AcceptBlock(*pblock, state, chainparams, &pindex, fRequested, dbp, &fNewBlock);
         if (pindex && pfrom) {
             mapBlockSource[pindex->GetBlockHash()] = pfrom->GetId();
+            if (fNewBlock) pfrom->nLastBlockTime = GetTime();
```

Comment:
It seems strange to me that the return value of AcceptBlock isn't checked before doing this. Doesn't seem like we would want to protect peers that are sending us invalid blocks.

Note that fNewBlock can be true even if CheckBlock/AcceptBlock return false.

Maybe there is a reason for this but it's not clear to me.